Add occ commands to enable and disable a user + a disabled user can n…

DeepDiver1975 opened the pull request:
…o longer login - fixes #23838 

@rullzer @LukasReschke @MorrisJobke

DeepDiver1975 committed:
Add occ commands to enable and disable a user + a disabled user can no longer login - fixes #23838

DeepDiver1975 committed:
Add provisioning api to enable and disable users

LukasReschke commented on apps/provisioning_api/lib/users.php:
```diff
@@ -333,6 +337,50 @@ public function deleteUser($parameters) {
+	 * @param bool $value
+	 * @return OC_OCS_Result
+	 */
+	private function setEnabled($parameters, $value) {
```
I'd like some integration tests on this covering the following scenarios:

1. Subadmin should be able to enable or disable an user in their group
2. Subadmins should NOT be able to enable or disable an user not in their group
3. Subadmins should not be able to disable users that have admin permissions even if they are in their group
4. Admins can enable and disable any user
5. Regular users will get an exception
6. It should not be possible to disable or enable ones own account, neither as admin nor as subadmin

cc @owncloud/qa

nickvergessen replied:
3. is not covered by the code?

LukasReschke replied:
It is. Check L370

LukasReschke replied:
Ah. Bloody GitHub, if I click on edit it shows "3." instead of "1."


Yes. 3, is not covered yet.

PVince81 replied:
@SergioBertolinSG do we have this covered now ?

SergioBertolinSG replied:
Yes it works fine.

DeepDiver1975 commented:
Additional integration tests to be implemented:
1. disable user
2. 503 is received via dav
3. 503 is received via ocs api calls
4. 503 is received via web requests

rullzer commented:
I did not check yet. But does this somehow yield the users shares invalid?

LukasReschke commented:
> I did not check yet. But does this somehow yield the users shares invalid?

Is this really required? I mean if you disable an user on an Active Directory server one can also still access files that belonged to them :see_no_evil:

butonic commented on lib/private/user/session.php:
```diff
@@ -229,7 +235,7 @@ public function login($uid, $password) {
 					}
 				} else {
-					return false;
+					throw new LoginException('User disabled');
```
Where do we localize the string so the user can be shown a proper message when he tries to login?

DeepDiver1975 replied:
Should be done here - thanks for the pointer

DeepDiver1975 committed:
Integration test for a disabled user

MorrisJobke commented on build/integration/features/webdav-related.feature:
```diff
@@ -241,3 +241,10 @@ Feature: webdav-related
 			| 1 |
 			| 3 |
+
+		Scenario: A disabled user cannot use webdav
```
This seems to fail 🙈

DeepDiver1975 commented:
:angry: 
```
Checked 4387 files in 45.2 seconds
Syntax error found in 1 file
------------------------------------------------------------
Parse error: ./lib/composer/composer/autoload_static.php:22
    20|         'OC\\Settings\\' => 
    21|         array (
  > 22|             0 => __DIR__ . '/../../..' . '/settings',
    23|         ),
    24|         'OC\\Core\\' => Unexpected '.', expecting ')'

DeepDiver1975 commented:
@nickvergessen @LukasReschke @MorrisJobke any idea why in this pr the composer generated file is causing trouble?

DeepDiver1975 commented:
issues seem to be related to php 5.4 and 5.5

DeepDiver1975 commented:
same fun in https://github.com/owncloud/core/pull/23708 - master broken?

LukasReschke commented:
Mhm. We don't ship any /lib/composer/composer/autoload_static.php, seems like composer on Travis generates this automatically… Let's see why it does that :wink:

LukasReschke commented:
Mhm… 

> Composer version 1.1-dev (4f0f8779cb6db67d2186e3480f5ca2243e9916e5) 2016-04-11 18:36:39

Caused by https://github.com/composer/composer/commit/fd2f51cea8e5f1ef978cd8f90b87b69dc1778976 likely… Let me see what one can do about that …

LukasReschke commented:
Ok. autoload_static.php is only loaded when PHP >= 5.6 is used. https://github.com/owncloud/core/pull/23935 adjusts the linter accordingly.

DeepDiver1975 commented:
@MorrisJobke @LukasReschke @rullzer please review and test once more - basic integration test added - needs further work @SergioBertolinSG 

THX

nickvergessen commented on apps/provisioning_api/tests/userstest.php:
```diff
@@ -59,7 +59,7 @@ protected function tearDown() {
 	}
 
 	protected function setup() {
```
camelCase missing as well

nickvergessen commented:
Btw in terms of tracking, would be nice to have a hook for the admin_audit app

DeepDiver1975 commented:
> Btw in terms of tracking, would be nice to have a hook for the admin_audit app

agreed - please open a ticket in core and enterprise so that we can consume it. THX

SergioBertolinSG commented on build/integration/features/bootstrap/Provisioning.php:
```diff
@@ -231,6 +230,20 @@ public function creatingTheGroup($group) {
 	/**
+	 * @When /^user "([^"]*)" is disabled$/
+	 */
+	public function userIsDisabled($user) {
```
I'd prefer this function to be named assureUserIsDisabled and @When /^Assure user "([^"]*)" is disabled$/

Because it actually disables the user, and a function for just checking will have this name.

DeepDiver1975 replied:
Looking at https://github.com/owncloud/core/pull/24071 - I can remove the integration test from this PR ?

@SergioBertolinSG

SergioBertolinSG replied:
Yes, that's fine I can add it there.

SergioBertolinSG replied:
Added, it can be removed from here.

DeepDiver1975 replied:
Why not bring all tests into this PR?

This way we can fix issues as they appear during tests being written?

SergioBertolinSG replied:
Ok, makes sense.

SergioBertolinSG committed:
Added integration tests directly in the code PR

SergioBertolinSG commented:
Current status:
- [x] Subadmin should be able to enable or disable an user in their group
- [x] Subadmins should NOT be able to enable or disable an user not in their group
- [x] Subadmins should not be able to disable users that have admin permissions even if they are in their group
- [ ] Admins can enable and disable any user
- [ ] Regular users will get an exception
- [ ] It should not be possible to disable or enable ones own account, neither as admin nor as subadmin
- [x] Disable user
- [x] 503 is received via dav
- [ ] 503 is received via ocs api calls
- [ ] 503 is received via web requests

SergioBertolinSG commented:
Sharing a file with a disabled user returns HTTP 200 instead of 503.

DeepDiver1975 commented on build/integration/features/sharing-v1.feature:
```diff
@@ -682,3 +682,15 @@ Feature: sharing
+      | path | welcome.txt |
+      | shareWith | user1 |
+      | shareType | 0 |
+    Then the HTTP status code should be "503"
```
Because this is OCS API v1 the http status code is always 200 - the OCS code should be 997/Unauthorized.
At least this is what the implementation is doing - we shall not break the current implementation because it will cause break in behavior - what we don't want to have.

SergioBertolinSG replied:
The answer is empty, not OCS code at all.

PVince81 replied:
@rullzer bug ?

rullzer replied:
most likely...

rullzer replied:
Ah I found it. So this is actually introduced by this PR. If a user is disabled the OCS Share API does not even enter the OCS part. It just errors out and prints the error page (hence the 503).

Let me push a fix for this

rullzer replied:
pushed. @SergioBertolinSG if you are using v2 the http status code should now be 401. And the OCS Status code should be 997

rullzer committed:
OCS API should catch LoginExceptions

Catching the login exception and returning false (login failed). Makes
the OCS API properly return data instead of printing the exception page.

rullzer commented:
So we also need additional logic in the shares.

1. userA shares a folder `foo` with userB
2. userA gets disabled
3. userB still had access to `foo`

Because else we can't use this for migration of home storages.

DeepDiver1975 commented:
> Because else we can't use this for migration of home storages.

I consider this topic for a second pr

rullzer commented:
> I consider this topic for a second pr

Ok great. just wanted to double check.

SergioBertolinSG committed:
Modified expected output according to rullzer

SergioBertolinSG commented:
There has emerged a couple of failures in  CommentsContext::teardownScenario() and TagsContext::tearDownScenario() .

DeepDiver1975 commented on lib/private/api.php:
```diff
@@ -356,7 +356,11 @@ private static function loginUser() {
+			try {
+				$return = OC_User::login($authUser, $authPw);
+			} catch (\OC\User\LoginException $e) {
+				return false;
```
we should log the exception from my understanding

PVince81 replied:
can you add it ? not sure if warning or debug ?

DeepDiver1975 commented:
Commit history need some cleanup .... will take care

PVince81 commented on tests/lib/user/session.php:
```diff
@@ -150,7 +165,12 @@ public function testLoginValidPasswordEnabled() {
+	 * @expectedException \OC\User\LoginException
+	 * @expectedExceptionMessage User disabled
+	 */
 	public function testLoginValidPasswordDisabled() {
```
oh, does it mean we already supported disabling users by setting an empty password ?

DeepDiver1975 replied:
No - what is meant: "Test the login of an disabled with a valid password" - the expectation is that the session is recreated

PVince81 commented on tests/lib/user/session.php:
```diff
@@ -347,6 +374,7 @@ public function testRememberLoginInvalidToken() {
 	}
 
 	public function testRememberLoginInvalidUser() {
```
I guess this covers "remember login token stops working after disabling user" ?

DeepDiver1975 replied:
yes - because the session is recreated

PVince81 commented:
Integration tests failing:
```
06:19:13     /ssd/jenkins/workspace/ocs-api-integration-tests-ci@3/build/integration/features/provisioning-v1.feature:503
06:19:13     /ssd/jenkins/workspace/ocs-api-integration-tests-ci@3/build/integration/features/sharing-v1.feature:700
```
(log is too big)

DeepDiver1975 commented:
Oh shit ... I guess I fucked up the rebase .....

PVince81 commented:
So much bad luck. Now that I was determined to review and test this, LDAP login is broken on this branch and master also... https://github.com/owncloud/core/issues/24409

PVince81 commented:
- [x] TEST: sync client stops syncing after disabling user (note: sync client shows **no error at all**, might need additional fixing there @dragotin)
- [x] TEST: disable local user
- [x] TEST: can still share with disabled user
- [x] TEST: can still access received local shares from disabled user
- [x] TEST: can still access received fed shares from disabled user
- [x] TEST: reenable user => sync client resumes without restart
- [x] TEST: delete file as recipient still moves file to disabled user's trashbin
- [x] TEST: edit file as recipient still creates versions for disabled user
- [x] TEST: encryption still works as recipient to disabled user's folder
- [x] TEST: web UI (ajax) reloads page when user disabled with open session
- [x] TEST: logout as disabled user
- [ ] TEST: disable LDAP user
- [ ] TEST: subadmin enable/disable

- [ ] BUG: CSRF mess up: 

1. Clear cookies
1. Login as "user1" (when enabled)
1. On the CLI, disable "user1"
1. In the web UI as "user1", click "Log out"
1. Login as "admin" again
=> "CSRF check failed" page.

DeepDiver1975 commented:
> Clear cookies
> Login as "user1" (when enabled)
> On the CLI, disable "user1"
> In the web UI as "user1", click "Log out"
> Login as "admin" again => "CSRF check failed" page.

@LukasReschke mind taking a look? THX

PVince81 commented on apps/provisioning_api/appinfo/routes.php:
```diff
@@ -41,6 +44,8 @@
 API::register('put', '/cloud/users/{userid}', [$users, 'editUser'], 'provisioning_api', API::USER_AUTH);
 API::register('delete', '/cloud/users/{userid}', [$users, 'deleteUser'], 'provisioning_api', API::SUBADMIN_AUTH);
+API::register('put', '/cloud/users/{userid}/enable', [$users, 'enableUser'], 'provisioning_api', API::SUBADMIN_AUTH);
+API::register('put', '/cloud/users/{userid}/disable', [$users, 'disableUser'], 'provisioning_api', API::SUBADMIN_AUTH);
```
In general to be more RESTful another idea would be to still do a "PUT" on "/cloud/users/{userid}" and set the "enabled" attribute to true or false. (PUT being an "edit" operation)

This also means that the GET operation would return the "enabled" flag as well with the user's properties.

I'm ok with keeping in that way if that's too much to rework.

PVince81 replied:
I see that GET already returns the flag. So intuitively as an API consumer, I'd be tempted to try and set it the way I described.

However now I see that it says "displayName" in the response but the PUT value is "display" :disappointed:

DeepDiver1975 replied:
Well - true restful would be to have only one GET to retrieve a json object holding all information and one PUT where we can push a full json object to store the new state.

But the whole provisioning api is not REST - so nothing to worry about - in the scope for v1.

PVince81 replied:
:see_no_evil: agreed

PVince81 commented:
Looks good from my POV :+1: 

@DeepDiver1975 did you test this with shib ?

DeepDiver1975 commented:
> @DeepDiver1975 did you test this with shib ?

no - I have currently no working shib test environment at hand - I suggest to merge and test this by QA

davitol commented:
Testing with Shib and Autoprovisioning:

Disabling/Enabling a user seems to work fine. But when disabling a shibboleth user and trying to log in with him, the following page is shown. Not sure if it is intended or not:


![screen shot 2016-05-03 at 13 56 44](https://cloud.githubusercontent.com/assets/10596560/14982507/26d9c550-1137-11e6-9822-565a994adadc.png)


No logs are written in owncloud.log

DeepDiver1975 commented:
I guess there is not much more we can do from a core pov. I guess we need to capture this situation and display a page like: you are disabled.

This has to be done in the shibboleth app

PVince81 commented:
"you are disabled" sounds a bit funny :smile:

davitol commented:
:+1: in that case

DeepDiver1975 commented:
@davitol please open an issue in the shibboleth repo - THX
